src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -1211,16 +1227,63 @@ namespace xclcpuemhal2 {
       return;
     }
 
+    setDeviceProcessStarted(false);
     std::string socketName = sock->get_name();
     if(socketName.empty() == false)// device is active if socketName is non-empty
     {
 #ifndef _WINDOWS
       xclClose_RPC_CALL(xclClose,this);
 #endif
     }
+   closemMessengerThread();
    saveDeviceProcessOutput();
   }
 
+  void CpuemShim::closemMessengerThread()
```

Comment:
```suggestion
  void CpuemShim::closeMessengerThread()
```
?
And actually I cannot see what you are closing here.
It looks more like you are waiting for the end of a thread.

Reply:
Updated the function name.